File: question/leetcode/leetcode0001.Two-Sum/leetcode0001.go
package leetcode

import "sort"

func twoSum(nums []int, target int) []int {
	// key: num, value: index
	table := make(map[int]int)

	for i, v := range nums {
		if p, ok := table[target-v]; ok {
			return []int{p, i}
		}
		table[v] = i
	}
	return nil
}

type Pair struct {
	Val int
	Idx int
}

func twoSumII(nums []int, target int) []int {
	n := len(nums)
	arr := make([]Pair, n)
	for i := 0; i < n; i++ {
		arr[i] = Pair{Val: nums[i], Idx: i}
	}
	sort.Slice(arr, func(i, j int) bool {
		return arr[i].Val < arr[j].Val
	})
	left, right := 0, n-1
	for left < right {
		sum := arr[left].Val + arr[right].Val
		if sum == target {
			return []int{arr[left].Idx, arr[right].Idx}
		} else if sum < target {
			left++
		} else {
			right--
		}
	}
	return nil
}

File: question/leetcode/leetcode0001.Two-Sum/leetcode0001.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

var twoSum = function (nums, target) {
	let map = new Map();
	for (let i = 0; i < nums.length; i++) {
		let index = map.get(target - nums[i]);
		if (index !== undefined) {
			return [i, index]
		}
		map.set(nums[i], i)
	}
	return [];
};

var twoSum = function (nums, target) {
	//题目要求返回原数组的索引
	const arr = nums.map((val, idx) => ({ val, idx })); // 记录原值和索引
	arr.sort((a, b) => a.val - b.val); // 按值排序j
	arr.sort()

	let left = 0;
	let right = arr.length - 1;

	while (left < right) {
		const sum = arr[left].val + arr[right].val;
		if (sum > target) {
			right--;
		} else if (sum < target) {
			left++;
		} else {
			return [arr[left].idx, arr[right].idx]; // 返回原始索引
		}
	}
	return [];
};